keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+
+- Request matching (similar to webhook Rules, NamespaceSelector,
+  ObjectSelector)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Automatic version conversion support (similar to webhooks MatchPolicy=Equivalent), with ability to introspect GVK of admitted resource in CEL expression.
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+severity, levels, failure policies, type safety, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  constraints:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    - expression: "object.spec.replicas <= config.maxReplicas"
+      # ...other rule related fields here...
+```
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` references the
+"configuration CRD" for this admission policy. For this example, the
+configuration CRD is `ReplicaLimit`.
+
+Note: This is a "Bring Your Own CRD" design. The admission policy definition
+author is responsible for providing the `ReplicaLimit` configuration CRD.
+
+`spec.constraints` specifies that the CEL expressions in
+`validations.expression` have access to the fields of the `v1.Deployment` kind.
+Multiple resources of the same kind may be matched, e.g. `*/scale`.
+
+To configure an admission policy, "policy configurations" of the configuration
+CRD kind are created. For example:
+
+```yaml
```

Comment:
What happens if I have two different ReplicaLimit objects that apply to a given namespace.  Say one selects on environment and the other selects on coolness-of-this-user?

Reply:
I'd imagine every ReplicaLimit object that matches the current object-under-test would be evaluated -- they'd all have to be satisfied for the request to be allowed.

Reply:
Correct, configurations can have overlapping match criteria. The behavior will be that the policy is evaluated for each matching configuration. For example, if you have a blocklist policy and two configurations (each containing their own set of blocklist entries) overlap, the policy is evaluated twice, once with each blocklist.  (We won't attempt to merge the blocklists, we'll just run the policy as-is with each blocklist).

Reply:
> Correct, configurations can have overlapping match criteria. The behavior will be that the policy is evaluated for each matching configuration. For example, if you have a blocklist policy and two configurations (each containing their own set of blocklist entries) overlap, the policy is evaluated twice, once with each blocklist. (We won't attempt to merge the blocklists, we'll just run the policy as-is with each blocklist).

Can you indicate that in this section?

Reply:
Done!